src/DSPC.sol

```diff
@@ -0,0 +1,290 @@
+    function drip() external;
+}
+
+interface ConvLike {
+    function btor(uint256 bps) external pure returns (uint256 ray);
+    function rtob(uint256 ray) external pure returns (uint256 bps);
+}
+
+/// @title Direct Stability Parameters Change Module
+/// @notice A module that allows direct changes to stability parameters with constraints
+/// @dev This contract manages stability parameters for ilks, DSR, and SSR with configurable limits
+/// @custom:authors [Oddaf]
+/// @custom:reviewers []
+/// @custom:auditors []
+/// @custom:bounties []
+contract DSPC {
+    // --- Structs ---
+    struct Cfg {
+        uint16 min; // Minimum rate in basis points
+        uint16 max; // Maximum rate in basis points
+        uint16 step; // Maximum rate change in basis points
+    }
+
+    struct ParamChange {
+        bytes32 id; // Identifier (ilk | "DSR" | "SSR")
+        uint256 bps; // New rate in basis points
+    }
+
+    // --- Immutables ---
+    /// @notice Stability fee rates
+    JugLike public immutable jug;
+    /// @notice DSR rate
+    PotLike public immutable pot;
+    /// @notice SSR rate
+    SUSDSLike public immutable susds;
+    /// @notice Rate conversion utility
+    ConvLike public immutable conv;
+
+    // --- Storage Variables ---
+    /// @notice Mapping of admin addresses
+    mapping(address => uint256) public wards;
+    /// @notice Mapping of addresses that can operate this module
+    mapping(address => uint256) public buds;
+    /// @notice Mapping of rate constraints
+    mapping(bytes32 => Cfg) private _cfgs;
+    /// @notice Circuit breaker flag
+    uint256 public bad;
+
+    // --- Events ---
+    /**
+     * @notice `usr` was granted admin access.
+     * @param usr The user address.
+     */
+    event Rely(address indexed usr);
+    /**
+     * @notice `usr` admin access was revoked.
+     * @param usr The user address.
+     */
+    event Deny(address indexed usr);
+    /**
+     * @notice `usr` was granted permission to change rates (call put()).
+     * @param usr The user address.
+     */
+    event Kiss(address indexed usr);
+    /**
+     * @notice Permission revoked for `usr` to change rates (call put()).
+     * @param usr The user address.
+     */
+    event Diss(address indexed usr);
+    /**
+     * @notice A contract parameter was updated.
+     * @param what The changed parameter name. ["bad"].
+     * @param data The new value of the parameter.
+     */
+    event File(bytes32 indexed what, uint256 data);
+    /**
+     * @notice A Ilk/DSR/SSR parameter was updated.
+     * @param id The Ilk/DSR/SSR identifier.
+     * @param what The changed parameter name. ["bad"].
+     * @param data The new value of the parameter.
+     */
+    event File(bytes32 indexed id, bytes32 indexed what, uint256 data);
+    /**
+     * @notice A batch of rate changes was executed.
+     * @param updates Array of rate updates.
+     */
+    event Put(ParamChange[] updates);
```

Comment:
Should match the function name:
```suggestion
    event Set(ParamChange[] updates);
```

I was also wondering if we should replace the batched payload by individual rates being set and emit multiple events, one per each update.
This way we can have proper indexing and discoverability of rates.
The counter-argument is that the changes are already logged in the underlying contract through the `File` event.